src/test/groovy/io/jsonwebtoken/codec/impl/Base64DecoderTest.groovy

```diff
@@ -0,0 +1,22 @@
+package io.jsonwebtoken.codec.impl
+
+import io.jsonwebtoken.lang.Strings
+import org.junit.Test
+
+import static org.junit.Assert.assertEquals
+
+class Base64DecoderTest {
+
+    @Test(expected = IllegalArgumentException)
+    void testDecodeWithNullArgument() {
+        new Base64Decoder().decode(null)
+    }
+
+    @Test
+    void testDecode() {
+        String encoded = 'SGVsbG8g5LiW55WM' // Hello 世界
```

Comment:
I would add in the the RFC test vectors also (I'm not sure where the best place to put them would be among the various encode/decode tests in this PR): https://tools.ietf.org/html/rfc4648#page-12

```
   BASE64("") = ""

   BASE64("f") = "Zg=="

   BASE64("fo") = "Zm8="

   BASE64("foo") = "Zm9v"

   BASE64("foob") = "Zm9vYg=="

   BASE64("fooba") = "Zm9vYmE="

   BASE64("foobar") = "Zm9vYmFy"
```

It would also be nice to cook up some examples to distinguish the differences between the url-safe one also.

Reply:
Code coverage reports 100% of Base64's implementation covered, for all statements and branches.

That said, definitely happy to add as many test vectors as we can reliably trust :)

Reply:
I wrote a quick program to generate printable ascii chars (it's a decent start at least, and we already have at least a few unicodes in the existing tests):

```
"special: [\n \t], ascii[32..126]: [ !\"#$%&'()*+,-./0123456789:;<=>?@ABCDEFGHIJKLMNOPQRSTUVWXYZ[\\]^_`abcdefghijklmnopqrstuvwxyz{|}~]\n"
```

I manually tinkered with it a little, so if my base64 encoding is off a little, it's probably me.  We should also dump a \r in there (I don't trust myself making a \r in vim and didn't write this out to a file in code like I probably should have).

Linux's base64 of it:

```
c3BlY2lhbDogWwogXHRdLCBhc2NpaVszMi4uMTI2XTogWyAhIiMkJSYnKCkqKywtLi8wMTIzNDU2Nzg5Ojs8PT4/QEFCQ0RFRkdISUpLTE1OT1BRUlNUVVZXWFlaW1xdXl9gYWJjZGVmZ2hpamtsbW5vcHFyc3R1dnd4eXp7fH1+XQo=
```

(again, subtle details might be wrong, so verify independent of my manual tinkering -- manual tinkering being where I added slashes b/c I printed the combination of ascii rather than writing to a file)

I'll write it to a file properly in a bit (going to eat lunch rq).

Reply:
@RyanBard I added the suggested test vectors (and duplicated them for Base64Url too).  PR has been updated.

Reply:
I'm reasonably confident in this one also:

```
input: "special: [\r\n \t], ascii[32..126]: [ !\"#$%&'()*+,-./0123456789:;<=>?@ABCDEFGHIJKLMNOPQRSTUVWXYZ[\\]^_`abcdefghijklmnopqrstuvwxyz{|}~]\n"
encoded: "c3BlY2lhbDogWw0KIAldLCBhc2NpaVszMi4uMTI2XTogWyAhIiMkJSYnKCkqKywtLi8wMTIzNDU2Nzg5Ojs8PT4/QEFCQ0RFRkdISUpLTE1OT1BRUlNUVVZXWFlaW1xdXl9gYWJjZGVmZ2hpamtsbW5vcHFyc3R1dnd4eXp7fH1+XQo="
```

I wrote that string to a file and used Linux's base64 util to encode (and then verified the decoding of that encoded str was my original output).

Reply:
@RyanBard in the block directly above, is everything to the right of `input: ` Java code? i.e. I can just copy-and-paste from that into a .java source code file?

Reply:
They were just string literals to dump into an assert, but since your tests are groovy and double quotes are GStrings, it required a little modification (and some replaces for the base64url).

I've PR'd into your branch here: https://github.com/jwtk/jjwt/pull/342

Scavenge what you need from it.